### DISCUSSION TITLE
django_idom app causes a significant delay

### DISCUSSION THREAD

**Discussion by numpde** [reactive-python/reactpy-django#128] [Problem] [answered]
When I do anything with Django's manage.py (migrate, collectstatic, runserver, ...), I see the two messages:
```console
2023-02-20T18:57:54+0000 | DEBUG | IDOM is in debug mode
IDOM has detected component sessions in the database, but no timestamp was found in cache. This may indicate that the cache has been cleared.
```
and it currently takes ~10sec. The DB is remote. 

Another scenario is when writing scratch code or tests that start with
```Python
import os
import django

os.environ.setdefault("DJANGO_SETTINGS_MODULE", "[myproject].settings")
django.setup()
```

Omitting `django_idom` from `INSTALLED_APPS` removes the delay.

django-idom==3.0.0a2
idom==1.0.0a3

**Comment by Archmonger** [comment]
Do you have a lot of Django template files?

**Reply by numpde** [reply]
No, and they display one component each, which are quite big, though, and some subcomponents use keys based on the underlying data.

 P.S.
```SQL
SELECT pg_size_pretty(pg_relation_size('django_idom_componentparams'));
```
gives 240 kB across 130 rows.

Could you please explain the purpose of this table (besides what's obvious from the table name)? And what would be the optimal type of DB to put it in?

**Reply by Archmonger** [reply]
That table is raw binary data containing all non-expired component sessions. Expired sessions get automatically purged during Django boot up (and also periodically during the runtime of the Django application)

I'm guessing your delays are related to the remote DB. It's fairly fast on local DBs.

If a 240kB database transfer + network operation is causing a 10+ second delay, this might be indicative of an underlying issue within your application/DB/network. 

However, within django-idom I could mask the issue by running the purge operation to run within a thread.

**Reply by rmorshea** [reply]
The table was introduced as part of [this change](https://github.com/idom-team/django-idom/pull/120) to allow arbitrary component params. We arrived at this approach after some discussion [here](https://github.com/idom-team/django-idom/issues/119). The table contents are described [here](https://github.com/idom-team/django-idom/blob/main/src/django_idom/models.py#L4-L7) where `data` contains the pickled component parameters.

I'm not familiar enough with Django to fully grasp what might be causing the problems here, but if database-backed component parameters are a problem, we could work on a solution that encrypts and embeds them in the templates as suggested [here](https://github.com/idom-team/django-idom/issues/119#issuecomment-1377720048). This would require some care to avoid any security vulnerabilities though, which is why we decided not to do this initially.

**Reply by numpde** [reply]
Can I choose which DB this is stored in? For example, during development, I think I'd prefer this locally and potentially in RAM.

This is perhaps answered [here](https://idom-team.github.io/django-idom/features/settings/).

**Reply by Archmonger** [reply]
As of right now, that's not a configurable setting. Good idea though, I will add that to the next version.

**Reply by numpde** [reply]
Thank you.

**Reply by Archmonger** [reply]
Will be resolved in this PR
- https://github.com/idom-team/django-idom/pull/127

**Reply by Archmonger** [reply, marked as answer]
`3.0.0a3` has been released. The ReactPy database backend is now configurable.
